tools/pal_inference.py

```diff
@@ -0,0 +1,471 @@
+
+        yield response
+        # stop when each sentence is finished, or if we exceed the maximum length
+        if unfinished_sequences.max() == 0 or stopping_criteria(
+                input_ids, scores):
+            break
+
+
+class GenericRuntime:
+    GLOBAL_DICT: dict = {}
+    LOCAL_DICT = None
+    HEADERS: List = []
+
+    def __init__(self):
+        self._global_vars = copy.copy(self.GLOBAL_DICT)
+        self._local_vars = copy.copy(
+            self.LOCAL_DICT) if self.LOCAL_DICT else None
+
+        for c in self.HEADERS:
+            self.exec_code(c)
+
+    def exec_code(self, code_piece: str) -> None:
+        exec(code_piece, self._global_vars)
+
+    def eval_code(self, expr: str) -> Any:
+        return eval(expr, self._global_vars)
+
+    def inject(self, var_dict: Dict[str, Any]) -> None:
+        for k, v in var_dict.items():
+            self._global_vars[k] = v
+
+    @property
+    def answer(self):
+        return self._global_vars['answer']
+
+
+class Timeout:
+
+    def __init__(self, seconds=1, error_message='Timeout'):
+        self.seconds = seconds
+        self.error_message = error_message
+
+    def timeout_handler(self, signum, frame):
+        raise TimeoutError(self.error_message)
+
+    def __enter__(self):
+        signal.signal(signal.SIGALRM, self.timeout_handler)
+        signal.alarm(self.seconds)
+
+    def __exit__(self, type, value, traceback):
+        signal.alarm(0)
+
+
+class ProgramInternLMInterface:
+
+    def __init__(
+        self,
+        model: AutoModelForCausalLM,
+        tokenizer: AutoTokenizer,
+        generation_config: GenerationConfig,
+        additional_eos_token_id: int = 103028,
+        get_answer_expr: str = 'solution()',
+        verbose: bool = False,
+    ):
+        """PAL interface wrap fun:`generate_interactive` to extract and execute
+        generated code.
+
+        Args:
+            model (AutoModelForCausalLM)
+            tokenizer (AutoTokenizer)
+            generation_config (GenerationConfig): Decode strategies
+            additional_eos_token_id (int): End of sentence token id, default: 103028
+            get_answer_expr (str): The function name of generated code, default: "solution()"
+            verbose (bool): Print error information
+        """
+        self.runtime = GenericRuntime()
+        self.history: List = []
+        self.model = model
+        self.tokenizer = tokenizer
+        self.generation_config = generation_config
+        self.additional_eos_token_id = additional_eos_token_id
+        self.answer_expr = get_answer_expr
+        self.verbose = verbose
+
+    def generate(self, prompt):
+        for cur_gen in generate_interactive(
+                model=self.model,
+                tokenizer=self.tokenizer,
+                prompt=prompt,
+                additional_eos_token_id=self.additional_eos_token_id,
+                **asdict(self.generation_config),
+        ):
+            continue
```

Comment:
what will this loop do? seems did not do anything.

Reply:
The `generate_interactive` function generate response word by word, we only need the last `cur_gen` as the final response.